### DISCUSSION TITLE
Compilation pains with iOS & ANEs.

### DISCUSSION THREAD

**Discussion by PhilCase** [airsdk/Adobe-Runtime-Support#3005] [General]
Hi,

I'm trying to build a new release with our app but hit problems when I try and test on iOS or deploy to Testflight.

Our app contains local ANEs and Distriqt ANEs.    (Onesignal Push Notifications, PDFReader, CameraUI, a custom VideoUtils and dependencies).   

To build the app I am using Windows 10 Pro and IntelliJ IDEA.   I'm using AIR 50.2.4.2.

At first it would not build for iOS so I upgraded all of the Distriqt ANEs using apm.   

I had problems with the "depth of the folder tree" during "apm update" - some files were not being installed as the file names would have been too long for windows - I fixed that.

I then had an app I could deploy locally but which crashed when it tried to start.    
I traced that down to a problem with the configuration/installation of the storyboard or the Assets.car - as suggested [here ](https://github.com/airsdk/Adobe-Runtime-Support/issues/2836).   At this point I had made a test app with the same name as my main app (using the same mobile provisioning file and p12 file - which proved that both the "mobile provisioning file" and "Apple Developer Licence" were okay).

I have made an iOS app for local testing (Windows 10 Pro - IntelliJ IDEA Ultimate, AIR 50.2.4.2).   It appears without an icon and fails because "its integrity cannot be verified".   I know my certificate is okay because I had made a test app that was using it for debugging.   (If I try adding Distriqt's CameraUI to my test app - it fails also.)

most recently I re-updated all of the ANEs, deployed it for debugging.    It created the "ipa" and gave me a message saying:
"Install iOS Application" 
WARNING: could not locate iTunesMetadata.plist in archive! 
WARNING: could not locate Payload/SmartCheck.app/SC_Info/SmartCheck.sinf in archive! 
ERROR: Install failed. Got error "ApplicationVerificationFailed" with code 0xe8008016: Failed to verify code signature of /var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.iKTfHJ/extracted/Payload/SmartCheck.app : 0xe8008016 (The executable was signed with invalid entitlements.)

I have tested uploading a release version to TestFlight .  That got rejected because the app was not signed with an appropriate distribution certificate (despite it being one that I just created).

So my question here is - do I need to test on a Mac and will I need to build my release version on a Mac to as suggested [here ](https://github.com/airsdk/Adobe-Runtime-Support/issues/2836) because of the Distriqt libraries - or is this not an issue?

Also - how do I get the logs from an iPhone X since "adt -deviceLog -platform ios -device(**deviceid**)  returns "-deviceLog is not supported for iOS" (when run on a Windows PC).

I'd be grateful for any help or pointers you can give please.

Kind regards

Phil

**Comment by FliplineStudios** [comment]
As far as I know, right now Windows can't correctly package for iOS since switching to LLVM as the mechanism for packaging, and the builds that it generates are rejected by Apple.  At the moment, only Macs can be used to package since they're using the traditional linker instead of LLVM, though I believe it requires a Mac which supports the newest Xcode and iOS SDKs and has them installed.

Andrew had mentioned somewhere that Apple finally published the open-source details for their latest linker, and that eventually they would be adapting this for AIR on Windows to build IPAs again like usual, but at the moment it's not available yet.

@ajwfrost Is there a timeline for when Windows will be able to package for iOS again?  I was actually going to open an issue about this today, as we're looking at releasing something within the next month or two, and our old Mac Mini that we use for uploading isn't compatible with the newer Xcode.  Trying to figure out if an update is coming soon where we can package on Windows, or if we'll need to purchase a newer Mac if we want to be able to package for iOS in the near future.

**Reply by ajwfrost** [reply]
Hi

The issue we're getting with Windows builds currently isn't that the binaries don't work on iPhones/iPads, it's purely an Apple App Store thing where the scanners that Apple use seem to think that we're using private APIs - which I believe is down to how the LLVM linker is different from the Apple LD64 linker. To address this, as mentioned we are planning to build using the Apple open source linker which has been updated with the necessary send-message support from Xcode 14, although I don't have a timeframe on that currently I'm afraid. For just local installations, or for ad-hoc or enterprise distribution, you should be okay still using Windows.

In terms of this issue though, it's sounding like a code-signing thing. 50.2.4.2 included changes to fix a problem that had been introduced by an update to Animate, and there may still be some bugs with that implementation - we know about one, where entitlements aren't properly DER-encoded if they include an array of elements... there's a patch for that available under https://github.com/airsdk/Adobe-Runtime-Support/issues/2989#issuecomment-1883531421

The error here is about code signatures and verification so it should be something we can resolve.. can you please try with above patch of ADT and if that still fails, if we can have your IPA we can analyse the code signature and see what's up. Even better if you're able to also build with 50.2.4.1 and (assuming that works) send both IPAs.. let me know if you need a transfer link.


thanks


---

> how do I get the logs from an iPhone X 

We don't have a way of doing this on Windows, in the same way that we can pick up logs from Android; on macOS you can use the "Console" app for it. But for now we're stuck with just whatever is available from Scout..

**Reply by PhilCase** [reply]
Thanks for the assist.

I've applied the patch to adt.jar.   I've copied this into the lib folder of the Flash 50.2.4.2 ADK.

The debug version on the app will compile (on the second try) and deploy to my iPhone X.   When I tap the icon it opens to a black screen that closes immediately.    

I've tried version 50.2.4.1 and version 50.2.3.8 and all have the same result.

**Reply by ajwfrost** [reply]
Hi
Are you able to get the crash log from the phone? via Settings -> Privacy & Security -> Analytics & Improvements -> Analytics Data...
I'm wondering, if this is iPhone X, whether it's the same issue as we just had reported where a symbol isn't being found in the right folder. Which is a potential problem across all the Windows builds, due to some recent(ish) updates in how Apple deploys the various symbol implementations in different libraries on different devices/OS versions...
thanks

**Reply by PhilCase** [reply]
Hi Andrew,

I've provided 3 startup reports and 2 "wakeups_resource" reports in the attached zip.  
Are these the files you wanted please?

[LOGS.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/13941493/LOGS.zip)

These are from a debug build.

thanks for the assist.

Phil.

**Reply by ajwfrost** [reply]
Hi

Thanks - looking at those though, it looks like it's missing an ANE library/framework?

```
  "exception" : {"codes":"0x0000000000000000, 0x0000000000000000","rawCodes":[0,0],"type":"EXC_CRASH","signal":"SIGABRT"},
  "termination" : {"code":1,"flags":518,"namespace":"DYLD","indicator":"Library missing"
  "details":["(terminated at launch; ignore backtrace)"],
  "reasons":["Library not loaded: @rpath\/OneSignalCore.framework\/OneSignalCore"
...
```

So the code signatures at least seem to check out, it's just that the IPA may be missing the "Frameworks" folder..? Is this from an ANE that you're creating yourself, or from a third party? It really seems to be missing, rather than having a missing dependency, which seems odd...

I would expect it to be found here:
```
tried: '\/private\/var\/containers\/Bundle\/Application\/F2712608-9D65-4318-AFD1-A60455B4F796\/SmartCheck.app\/Frameworks\/OneSignalCore.framework\/OneSignalCore' (no such file)
```

Are you able to unzip your IPA file and check whether it's in there under `Payloads/SmartCheck.app/Frameworks`?

thanks

**Reply by marchbold** [reply]
@PhilCase If you are using our ANEs and `apm` then check you are adding the `assets/ios/Frameworks` folder (or `assets/ios_simulator/Frameworks` for simulator builds) to your build. 

This folder must be added otherwise you'll get a crash on launch when the app is attempting to find the dynamic frameworks.

(More details here: https://docs.airnativeextensions.com/docs/pushnotifications/onesignal/add-the-extension)

**Reply by PhilCase** [reply]
Thanks @ajwfrost + @marchbold will try.

I'm appreciating that Andrew was able to spot that so quickly. 👍

**Reply by ajwfrost** [reply]
I'm wondering whether there's a way we can detect this at compilation time ... @marchbold just wondering, I would assume that our linker command includes a list of the frameworks that are required here, but that on its own may not help since some frameworks are system ones that we don't need in the package... but maybe, if we can check whether a framework is available under the SDK folder, we assume we don't need it, but if we don't find it there, then we could flag a warning or error to say that the framework isn't being packaged in the IPA?

Or any other thoughts on whether we can somehow capture where a framework is a dependency both for the linker command and for the package's Frameworks folder?

thanks

**Reply by PhilCase** [reply]
Hi - thanks @ajwfrost + @marchbold that is what it was.  

The Distriqt OneSignal ANE was missing the extra Frameworks folder in the root of the iOS project.   I was not aware of this since I've upgraded from an earlier version of the library.

I am able to make an iOS debug version.    I'll try and make an iOS release version at some point.

**Reply by marchbold** [reply]
@PhilCase Glad to hear!! 

@ajwfrost I suppose it comes down to if you can distinguish between a dynamic and a static frameworks easily. You'd just have to go through the list of packaged framework dependencies and then scan them to determine if they are static or dynamic and then check the Frameworks folder for any dynamic ones.

As an idea, we could flag them as dynamic in the ANE and have AIR add them automatically to the `Frameworks` folder eg

```xml
<packagedDependency type="dynamic">OneSignalOSCore.framework</packagedDependency>
```

But as it stands with `apm` it just requires people to add the `assets/ios/*` contents to the root of their app, this includes any assets/bundles/files/dynamic frameworks that SDKs in the ANEs use, you could go this path and look for that directory to add automatically?
